securesystemslib/signer/_signer.py

```diff
@@ -0,0 +1,264 @@
+
+        signer = SIGNER_FOR_URI_SCHEME[scheme]
+        return signer.from_priv_key_uri(
+            priv_key_uri, public_key, secrets_handler
+        )
+
+
+class SSlibSigner(Signer):
+    """A securesystemslib signer implementation.
+
+    Provides a sign method to generate a cryptographic signature with a
+    securesystemslib-style rsa, ed25519 or ecdsa key. See keys module
+    for the supported types, schemes and hash algorithms.
+
+    SSlibSigners should be instantiated with Signer.from_priv_key_uri().
+    These private key URI schemes are supported:
+    * "envvar:<VAR>":
+        VAR is an environment variable with unencrypted private key content.
+           envvar:MYPRIVKEY
+    * "file:<PATH>?encrypted=[true|false]":
+        PATH is a file path to a file with unencrypted private key content.
+           file:path/to/file?encrypted=true
+           file:/abs/path/to/file?encrypted=false
+
+    Attributes:
+        key_dict:
+            A securesystemslib-style key dictionary. This is an implementation
+            detail, not part of public API
+    """
+
+    ENVVAR_URI_SCHEME = "envvar"
+    FILE_URI_SCHEME = "file"
+
+    def __init__(self, key_dict: Dict):
+        self.key_dict = key_dict
+
+    @classmethod
+    def from_priv_key_uri(
+        cls,
+        priv_key_uri: str,
+        public_key: Key,
+        secrets_handler: Optional[SecretsHandler] = None,
+    ) -> "SSlibSigner":
+        """Constructor for Signer to call
+
+        Arguments:
+            priv_key_uri: private key URI described in class doc
+            public_key: Key object.
+
+        Raises:
+            OSError: Reading the file failed with "file:" URI
+            ValueError: URI is unsupported or environment variable was not set
+                with "envvar:" URIs
+
+        Returns:
+            SSlibSigner for the given private key URI.
+        """
+        if not isinstance(public_key, SSlibKey):
+            raise ValueError(f"Expected SSlibKey for {priv_key_uri}")
+
+        uri = parse.urlparse(priv_key_uri)
+
+        if uri.scheme == cls.ENVVAR_URI_SCHEME:
+            # read private key from environment variable
+            private = os.getenv(uri.path)
+            if private is None:
+                raise ValueError(f"Unset env var for {priv_key_uri}")
+
+        elif uri.scheme == cls.FILE_URI_SCHEME:
+            params = dict(parse.parse_qsl(uri.query))
+            if "encrypted" not in params:
+                raise ValueError(f"{uri.scheme} requires 'encrypted' parameter")
+
+            # read private key (may be encrypted or not) from file
+            with open(uri.path, "rb") as f:
+                private = f.read().decode()
+
+            if params["encrypted"] != "false":
+                if not secrets_handler:
+                    raise ValueError("encrypted key requires a secrets handler")
+
+                secret = secrets_handler("passphrase")
```

Comment:
Can you explain how `secrets_handler` is supposed to be used? I don't understand why it can be passed a name for the secret, and why this is hardcoded `"passphrase"` here.

Reply:
* signer calls the method when it needs a secret from user
* signer providers the name of the secret (just in case there could be several secrets -- this is a bit theoretical)
* the application implements SecretsHandler: here's a simple one for a CLI app:
```python
from getpass import getpass

def my_secret_handler(secret_name:str ) -> str:
    return getpass(f"Please provide {secret_name}: ")
```

Reply:
oh, and if it's not clear: secrets_handler is a method (and not a string argument to from_priv_key_uri()) because
* we don't know if or when the secret is needed: there could be cases where it's requested at signing time and sometimes it might not be requested at all (think yubikey with a short "unlock window" after first PIN input)
* we don't necessarily know which secret is required (again, this is a bit theoretical but possible)

Reply:
Ah, now I understand, thanks for explaining! I was confused by the comment above the secrets handler definition: "The argument is the name of the secret", when really it can be anything you implement in the secrets handler. That plus the hardcoded "passphrase", which might as well be hardcoded into the secrets handler. But I guess it makes sense for some use cases. There might be cases where no arg or multiple or non-string args might also make sense.

Reply:
Okay, but if the passed name of the secret is not really a canonical identifier and not documented. How should the application, which provides the handler know what to expect and how to deal with it?